util/src/main/scala/Money.scala

```diff
@@ -8,6 +8,43 @@ import cats.Monoid
 
 import scala.math._
 import BigDecimal.RoundingMode._
+import scala.math.BigDecimal.RoundingMode
+
+sealed trait BaseMoney {
+  def `type`: String
+
+  def currency: Currency
+
+  // Use with CAUTION! will loose precision in case of a high precision money value
+  def centAmount: Long
+
+  def fractionDigits: Int
+
+  def toMoneyWithPrecisionLoss: Money
+
+  def + (m: Money)(implicit mode: RoundingMode): BaseMoney
+  def + (m: HighPrecisionMoney)(implicit mode: RoundingMode): BaseMoney
+  def + (m: BaseMoney)(implicit mode: RoundingMode): BaseMoney
```

Comment:
In this case these are just abstract methods here. If we implement these in Money and HPMoney, we can also narrow down the return type (as you also mentioned in the other comment). At least for HPMoney. For example, this clode:

```scala
hpMoney + someBaseMoney
```

would be of type (for scalac) HPMoney instead of just BaseMoney.

Does it answer the question or did you mean something else?

Reply:
But can't you also do that if you only declare `def + (BaseMoney): BaseMoney` here?

I don't see the value of declaring the narrowed-down types for the argument, if the return type is always `BaseMoney`.

```scala
trait BaseMoney[ConcreteMoney <: BaseMoney] {
  def + (m: Money)(implicit mode: RoundingMode): BaseMoney
  def + (m: ConcreteMoney)(implicit mode: RoundingMode): ConcreteMoney
}
```

The above would provide at least some value, as it would force an implementation to return it's own type if added with itself... but it's probably not worth the overhead.

Reply:
We may could use `this.type` here.

Reply:
@yanns I don't think that `this.type` would be accurate here since `money + hpMoney = hpMoney`. `this.type` would imply that the result must be of `Money` type.

Reply:
Do we want to support `money + hpMoney`?
I'd have guess we only support operations on similar types.

Reply:
@cneijenhuis yeah, I also was thinking about introducing f-bounded type, but for such small use-case it would be overkill and add a bit more frustration when used elsewhere in the code. 

I guess these overloads will provide a some convenience. I think we added some of these as we were refactoring the backed codebase. Overloads + narrowing down the return type in a subtype adds a small advantage and some safety net. Just an example:

* `money + hpMoney` - the type of the whole expression is `HPMoney` (at compile time)
* `money + money` - the type of the whole expression is `Money` (at compile time)
* `money + baseMoney` - the type of the whole expression is `BaseMoney` (at compile time)

Reply:
@yanns I think we need support for `money + hpMoney` (either conveniently supported here directly or done elsewhere with external money → hpMoney conversion). The operation itself safe and does not lose precision.